halo2_proofs/src/plonk/circuit.rs

```diff
@@ -1545,6 +1685,472 @@ impl<F: Field> Gate<F> {
+                    column_index: query.column_index,
+                    rotation: query.rotation,
+                })
+            }
+            ExpressionMid::Challenge(c) => Expression::Challenge(*c),
+            ExpressionMid::Negated(e) => Expression::Negated(Box::new(self.as_expression(e))),
+            ExpressionMid::Sum(lhs, rhs) => Expression::Sum(
+                Box::new(self.as_expression(lhs)),
+                Box::new(self.as_expression(rhs)),
+            ),
+            ExpressionMid::Product(lhs, rhs) => Expression::Product(
+                Box::new(self.as_expression(lhs)),
+                Box::new(self.as_expression(rhs)),
+            ),
+            ExpressionMid::Scaled(e, c) => Expression::Scaled(Box::new(self.as_expression(e)), *c),
+        }
+    }
+}
+
+/// This is a description of the circuit environment, such as the gate, column and
+/// permutation arrangements.
+#[derive(Debug, Clone)]
+pub struct ConstraintSystemV2Backend<F: Field> {
+    pub(crate) num_fixed_columns: usize,
+    pub(crate) num_advice_columns: usize,
+    pub(crate) num_instance_columns: usize,
+    pub(crate) num_challenges: usize,
+
+    /// Contains the index of each advice column that is left unblinded.
+    pub(crate) unblinded_advice_columns: Vec<usize>,
+
+    /// Contains the phase for each advice column. Should have same length as num_advice_columns.
+    pub(crate) advice_column_phase: Vec<u8>,
+    /// Contains the phase for each challenge. Should have same length as num_challenges.
+    pub(crate) challenge_phase: Vec<u8>,
+
+    pub(crate) gates: Vec<GateV2Backend<F>>,
+
+    // Permutation argument for performing equality constraints
+    pub(crate) permutation: permutation::Argument,
+
+    // Vector of lookup arguments, where each corresponds to a sequence of
+    // input expressions and a sequence of table expressions involved in the lookup.
+    pub(crate) lookups: Vec<lookup::ArgumentV2<F>>,
+
+    // Vector of shuffle arguments, where each corresponds to a sequence of
+    // input expressions and a sequence of shuffle expressions involved in the shuffle.
+    pub(crate) shuffles: Vec<shuffle::ArgumentV2<F>>,
+
+    // List of indexes of Fixed columns which are associated to a circuit-general Column tied to their annotation.
+    pub(crate) general_column_annotations: HashMap<metadata::Column, String>,
+}
+
+/// Witness calculator.  Frontend function
+#[derive(Debug)]
+pub struct WitnessCalculator<'a, F: Field, ConcreteCircuit: Circuit<F>> {
+    k: u32,
+    n: usize,
+    unusable_rows_start: usize,
+    circuit: &'a ConcreteCircuit,
+    config: &'a ConcreteCircuit::Config,
+    cs: &'a ConstraintSystem<F>,
+    instances: &'a [&'a [F]],
+    next_phase: u8,
+}
+
+impl<'a, F: Field, ConcreteCircuit: Circuit<F>> WitnessCalculator<'a, F, ConcreteCircuit> {
+    /// Create a new WitnessCalculator
+    pub fn new(
+        k: u32,
+        circuit: &'a ConcreteCircuit,
+        config: &'a ConcreteCircuit::Config,
```

Comment:
We could try to make a fn for `ConcreteCircuit` which returns a ref to the config. 

Such that we pass 1 less argument.

Reply:
I'm not sure I follow you.  `ConcreteCircuit` is a trait which builds the config, but doesn't hold a reference to it.

You could create the config again, but then you're doing repeated work because in the setup you've already done that.

Reply:
Hence why I think we could have a getter and assume that the circuit always holds inside it's own built config.

Reply:
Following original halo2 API, the circuit type only holds the inputs to the circuit, not the configuration.  To keep the same functionality, I would avoid doing any change in this PR, but we can discuss it in a new issue.

Nevertheless I'm not convinced about the benefits of this change, I think the current design regarding that part is fine.